src/pattern-acceptor.js

```diff
@@ -0,0 +1,530 @@
+  }
+}
+
+export const acceptRegex = (pattern, unicode) => {
+  let state = new PatternAcceptorState(pattern, unicode);
+  let accepted = acceptDisjunction(state);
+  if (accepted.matched && state.unicode) {
+    for (let backreference of state.backreferences) {
+      if (backreference > state.capturingGroups) {
+        return false;
+      }
+    }
+  }
+  return accepted.matched;
+};
+
+const backtrackOnFailure = func => state => {
+  let savedIndex = state.index;
+  let oldBackreferences = state.backreferences.slice(0);
+  let oldCapturingGroups = state.capturingGroups;
+  let val = func(state);
+  if (!val.matched) {
+    state.index = savedIndex;
+    state.backreferences = oldBackreferences;
+    state.capturingGroups = oldCapturingGroups;
+  }
+  return val;
+};
+
+const acceptUnicodeEscape = backtrackOnFailure(state => {
+  if (!state.eat('u')) {
+    return { matched: false };
+  }
+  if (state.unicode && state.eat('{')) {
+    let digits = [];
+    while (!state.eat('}')) {
+      let digit = state.eatAny(...hexDigits);
+      if (digit === null) {
+        return { matched: false };
+      }
+      digits.push(digit);
+    }
+    let value = parseInt(digits.join(''), 16);
+    return value > 0x10FFFF ? { matched: false } : { matched: true, value };
+  }
+  let digits = [0, 0, 0, 0].map(() => state.eatAny(...hexDigits));
+  if (digits.some(digit => digit === null)) {
+    return { matched: false };
+  }
+  let value = parseInt(digits.join(''), 16);
+  if (state.unicode && value >= 0xD800 && value <= 0xDBFF) {
+    let surrogatePairValue = backtrackOnFailure(subState => {
+      if (!subState.eat('\\u')) {
+        return { matched: false };
+      }
+      let digits2 = [0, 0, 0, 0].map(() => subState.eatAny(...hexDigits));
+      if (digits2.some(digit => digit === null)) {
+        return { matched: false };
+      }
+      let value2 = parseInt(digits2.join(''), 16);
+      if (value2 < 0xDC00 || value2 >= 0xE000) {
+        return { matched: false };
+      }
+      return { matched: true, value: 0x10000 + ((value & 0x03FF) << 10) + (value2 & 0x03FF) };
+    })(state);
+    if (surrogatePairValue.matched) {
+      return surrogatePairValue;
+    }
+  }
+  return { matched: true, value };
+});
+
+const acceptDisjunction = (state, terminator) => {
+  do {
+    if (terminator !== void 0 && state.eat(terminator)) {
+      return { matched: true };
+    } else if (state.match('|')) {
+      continue;
+    }
+    if (!acceptAlternative(state, terminator).matched) {
+      return { matched: false };
+    }
+  } while (state.eat('|'));
+  return { matched: terminator === void 0 || !!state.eat(terminator) };
+};
+
+const acceptAlternative = (state, terminator) => {
+  while (!state.match('|') && !state.empty() && (terminator === void 0 || !state.match(terminator))) {
+    if (!acceptTerm(state).matched) {
+      return { matched: false };
+    }
+  }
+  return { matched: true };
+};
+
+const orMatched = (...predicates) => state => {
```

Comment:
Can we rename this to `any` or `anyOf` or something? Those names are more traditional.

Also rename `predicates` to `acceptors`; `predicate` implies returning only `true` or `false`, which these do not.